anycloud/node/src/datastore.ts

```diff
@@ -0,0 +1,90 @@
+import fetch from 'node-fetch';
+const https = require('https');
+
+export class DataStore {
+  private localDS: any;
+  private clusterSecret?: string;
+  private isLocal: boolean;
+  private headers?: any;
+  private ctrlPortUrl: string;
+  private ctrlPortAgent: any;
+
+  constructor() {
+    this.localDS = {};
+    this.clusterSecret = process.env.CLUSTER_SECRET;
+    this.isLocal = !this.clusterSecret;
+    this.headers = this.clusterSecret ? { [this.clusterSecret]: 'true' } : null;
+    this.ctrlPortUrl = 'https://localhost:4142/app/';
+    // To avoid failure due to self signed certs
+    this.ctrlPortAgent = new https.Agent({
+      rejectUnauthorized: false,
+    });
+  }
+
+  private async request(method: string, url: string, body?: any): Promise<string> {
+    const response = await fetch(url, {
+      agent: this.ctrlPortAgent,
+      method: method,
+      headers: this.headers,
+      body,
+    });
+    return await response.text();
+  }
+  async get(dsKey: string): Promise<string> {
+    if (!dsKey) return undefined;
+    if (this.isLocal) {
+      if (dsKey in this.localDS) {
+        const val = this.localDS[dsKey]; 
+        if (typeof(val) === 'object') {
+          return JSON.stringify(val);
+        } else {
+          return `${val}`;
+        }
+      } else {
+        return undefined;
+      }
+    }
+    try {
+      return await this.request('GET', `${this.ctrlPortUrl}get/${dsKey.toString()}`);
+    } catch (e) {
+      return e;
+    }
+  }
+
+  async set(dsKey: string, dsValue: any): Promise<string> {
+    if (!dsKey) return 'fail';
+    if (this.isLocal) {
+      this.localDS[dsKey] = dsValue;
+      return 'ok';
+    }
+    try {
+      return await this.request('POST', `${this.ctrlPortUrl}set/${dsKey.toString()}`, dsValue);
+    } catch (_) {
+      return 'fail';
+    }
+  }
+
+  async del(dsKey: string): Promise<boolean> {
+    if (!dsKey) return false;
+    if (this.isLocal) {
+      return dsKey in this.localDS ? delete this.localDS[dsKey] : false;
+    }
+    try {
+      return await this.request('GET', `${this.ctrlPortUrl}del/${dsKey.toString()}`) === 'true';
```

Comment:
whyyyyyy does `delete <symbol>` return `true` not only if the key already existed, but if the key didn't exist?

Reply:
why does it return `true` at all????

Reply:
![Screenshot from 2021-06-03 11-13-28](https://user-images.githubusercontent.com/765551/120692567-c967f200-c45c-11eb-8171-2da99d0f7c46.png)

lolwat.

Though I think I know why.

![Screenshot from 2021-06-03 11-15-56](https://user-images.githubusercontent.com/765551/120692802-1c41a980-c45d-11eb-9fb4-2b54e773b38d.png)

You can define undeleteable properties on objects, so the boolean indicates "does this property no longer exist".

Reply:
interesting... still a very weird decision and i don't like it, nope, not one bit